edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/metrics/AvaliabilityMetrics.cs

```diff
@@ -0,0 +1,155 @@
+            new List<string> { "module_name", "module_version" });
+
+        private static readonly IMetricsGauge WeeklyAvailability = Util.Metrics.Metrics.Instance.CreateGauge(
+            "weekly_availability",
+            "total availability for the last 7 days",
+            new List<string> { "module_name", "module_version" });
+
+        public static readonly ILogger Log = Logger.Factory.CreateLogger<Availability>();
+        public static ISystemTime Time = SystemTime.Instance;
+        public static Option<string> StoragePath = Option.None<string>();
+
+        private static Lazy<List<(Availability lifetime, WeeklyAvailability weekly)>> availabilities = new Lazy<List<(Availability lifetime, WeeklyAvailability weekly)>>(LoadData);
+
+        static AvailabilityMetrics()
+        {
+            AppDomain.CurrentDomain.ProcessExit += SaveData;
+        }
+
+        public static void ComputeAvailability(ModuleSet desired, ModuleSet current)
+        {
+            /* Get all modules that are not running but should be */
+            var down = new HashSet<string>(current.Modules.Values
+                .Where(c =>
+                    (c is IRuntimeModule) &&
+                    (c as IRuntimeModule).RuntimeStatus != ModuleStatus.Running &&
+                    desired.Modules.TryGetValue(c.Name, out var d) &&
+                    d.DesiredStatus == ModuleStatus.Running)
+                .Select(c => c.Name));
+
+            /* Get all correctly running modules */
+            var up = new HashSet<string>(current.Modules.Values
+                .Where(c => (c is IRuntimeModule) && (c as IRuntimeModule).RuntimeStatus == ModuleStatus.Running)
+                .Select(c => c.Name));
+
+            /* Add points for all modules found */
+            foreach ((Availability lifetime, WeeklyAvailability weekly) in availabilities.Value)
+            {
+                string name = lifetime.Name;
+                if (down.Remove(name))
+                {
+                    lifetime.AddPoint(false);
+                    weekly.AddPoint(false);
+                }
+                else if (up.Remove(name))
+                {
+                    lifetime.AddPoint(true);
+                    weekly.AddPoint(true);
+                }
+                else
+                {
+                    /* stop calculating if in stopped state or not deployed */
+                    lifetime.NoPoint();
+                    weekly.NoPoint();
+                }
+
+                LifetimeAvailability.Set(lifetime.AvailabilityRatio, new[] { lifetime.Name, lifetime.Version });
+                WeeklyAvailability.Set(weekly.AvailabilityRatio, new[] { weekly.Name, weekly.Version });
+            }
+
+            /* Add new modules to track */
+            foreach (string module in down.Union(up))
+            {
+                availabilities.Value.Add((new Availability(module, "tempNoVersion", Time), new WeeklyAvailability(module, "tempNoVersion", Time)));
```

Comment:
`"tempNoVersion"`?!?

Reply:
module versions aren't known in edge agent rn. will either figure out how to get them or remove this tag.